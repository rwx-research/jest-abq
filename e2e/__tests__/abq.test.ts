/**
 * Copyright (c) ReadWriteExecute, Inc. and its affiliates. All Rights Reserved.
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {ChildProcess, spawn} from 'child_process';
import * as path from 'path';
import {
  filterManifestForSnapshot,
  filterTestResultForSnapshot,
  pathForAbqTestFile,
  runAbqJest,
} from '../abqUtils';

// set this to true to see the server output with your jest results
const DEBUG_SERVER = false;

let serverToCleanup: ChildProcess;

const stdioVal = DEBUG_SERVER ? 'inherit' : 'ignore';

function spawnServer(
  args: Array<any> = [],
): Promise<[string, () => Promise<Array<any>>]> {
  const serverMessages: Array<any> = [];
  let resolveMessages: any;
  const messagesPromise = new Promise<Array<any>>(resolve => {
    resolveMessages = resolve;
  });
  async function getMessages() {
    return messagesPromise;
  }
  return new Promise(resolve => {
    const serverPath = path.resolve(__dirname, '../server.js');
    const server = (serverToCleanup = spawn(
      'node',
      [serverPath, ...args.map(a => JSON.stringify(a))],
      {
        stdio: [stdioVal, stdioVal, stdioVal, 'ipc'],
      },
    ));

    server.on('message', message => {
      if (message.socketString) {
        resolve([message.socketString, getMessages]);
      } else if (message === '--CONNECTION CLOSED--') {
        resolveMessages(serverMessages);
      } else {
        serverMessages.push(message);
      }
    });
  });
}

function ctest(...args: Parameters<typeof test>) {
  if (process.env.JEST_JASMINE === '1') {
    // ABQ does not support jest-jasmine2
    test.skip(...args);
    return;
  }
  test(...args);
}

afterEach(async () => {
  if (serverToCleanup) {
    serverToCleanup.kill();
  }
});

ctest('ABQ_GENERATE_MANIFEST sends the manifest to the socket', async () => {
  expect.assertions(1);

  const [socketString, getMessages] = await spawnServer();

  await runAbqJest(
    {
      ABQ_GENERATE_MANIFEST: '1',
      ABQ_SOCKET: socketString,
    },
    async () => {
      const serverMessages = (await getMessages()).map(
        filterManifestForSnapshot,
      );
      expect(serverMessages).toMatchSnapshot();
    },
  );
});

ctest('ABQ_SOCKET runs Jest in ABQ mode', async () => {
  expect.assertions(1);

  const [socketString, getMessages] = await spawnServer([
    {
      id: pathForAbqTestFile('sum.test.js'),
      meta: {
        fileName: pathForAbqTestFile('sum.test.js'),
      },
      tags: [],
      type: 'test',
    },
    {
      id: pathForAbqTestFile('failing.test.js'),
      meta: {
        fileName: pathForAbqTestFile('failing.test.js'),
      },
      tags: [],
      type: 'test',
    },
  ]);

  await runAbqJest(
    {
      ABQ_SOCKET: socketString,
    },
    async () => {
      const serverMessages = (await getMessages()).map(
        filterTestResultForSnapshot,
      );
      expect(serverMessages).toMatchSnapshot();
    },
  );
});

ctest('Reports all tests in a file', async () => {
  expect.assertions(1);

  const [socketString, getMessages] = await spawnServer([
    {
      id: pathForAbqTestFile('sum.test.js'),
      meta: {
        fileName: pathForAbqTestFile('sum.test.js'),
      },
      tags: [],
      type: 'test',
    },
  ]);

  await runAbqJest(
    {
      ABQ_SOCKET: socketString,
    },
    async () => {
      const serverMessages = (await getMessages()).map(
        filterTestResultForSnapshot,
      );
      expect(serverMessages).toMatchSnapshot();
    },
  );
});

ctest(
  'ABQ_HIDE_NATIVE_OUTPUT hides all output in the jest process',
  async () => {
    expect.assertions(2);

    const [socketString, getMessages] = await spawnServer([
      {
        id: pathForAbqTestFile('sum.test.js'),
        meta: {
          fileName: pathForAbqTestFile('sum.test.js'),
        },
        tags: [],
        type: 'test',
      },
      {
        id: pathForAbqTestFile('failing.test.js'),
        meta: {
          fileName: pathForAbqTestFile('failing.test.js'),
        },
        tags: [],
        type: 'test',
      },
    ]);

    const {stderr, stdout} = await runAbqJest(
      {
        ABQ_HIDE_NATIVE_OUTPUT: '1',
        ABQ_SOCKET: socketString,
      },
      getMessages,
    );

    expect(stderr).toBe('');
    expect(stdout).toBe('');
  },
);

ctest('ABQ mode handles errors in a test', async () => {
  expect.assertions(1);

  const [socketString, getMessages] = await spawnServer([
    {
      id: pathForAbqTestFile('errors.test.js'),
      meta: {
        fileName: pathForAbqTestFile('errors.test.js'),
      },
      tags: [],
      type: 'test',
    },
  ]);

  await runAbqJest(
    {
      ABQ_SOCKET: socketString,
    },
    async () => {
      const serverMessages = (await getMessages()).map(
        filterTestResultForSnapshot,
      );
      expect(serverMessages).toMatchSnapshot();
    },
  );
});

ctest('ABQ mode handles errors outside of test execution', async () => {
  expect.assertions(1);

  const [socketString, getMessages] = await spawnServer([
    {
      id: pathForAbqTestFile('error_setup.test.js'),
      meta: {
        fileName: pathForAbqTestFile('error_setup.test.js'),
      },
      tags: [],
      type: 'test',
    },
  ]);

  await runAbqJest(
    {
      ABQ_SOCKET: socketString,
    },
    async () => {
      const serverMessages = (await getMessages()).map(
        filterTestResultForSnapshot,
      );
      expect(serverMessages).toMatchSnapshot();
    },
  );
});

ctest('ABQ mode handles skipped tests', async () => {
  expect.assertions(1);

  const [socketString, getMessages] = await spawnServer([
    {
      id: pathForAbqTestFile('skip.test.js'),
      meta: {
        fileName: pathForAbqTestFile('skip.test.js'),
      },
      tags: [],
      type: 'test',
    },
  ]);

  await runAbqJest(
    {
      ABQ_SOCKET: socketString,
    },
    async () => {
      const serverMessages = (await getMessages()).map(
        filterTestResultForSnapshot,
      );
      expect(serverMessages).toMatchSnapshot();
    },
  );
});

ctest('ABQ mode handles todo tests', async () => {
  expect.assertions(1);

  const [socketString, getMessages] = await spawnServer([
    {
      id: pathForAbqTestFile('todo.test.js'),
      meta: {
        fileName: pathForAbqTestFile('todo.test.js'),
      },
      tags: [],
      type: 'test',
    },
  ]);

  await runAbqJest(
    {
      ABQ_SOCKET: socketString,
    },
    async () => {
      const serverMessages = (await getMessages()).map(
        filterTestResultForSnapshot,
      );
      expect(serverMessages).toMatchSnapshot();
    },
  );
});

ctest('ABQ mode handles ID generation for tests in loops', async () => {
  expect.assertions(4);

  const [socketString, getMessages] = await spawnServer([
    {
      id: pathForAbqTestFile('looped.test.js'),
      meta: {
        fileName: pathForAbqTestFile('looped.test.js'),
      },
      tags: [],
      type: 'test',
    },
  ]);

  await runAbqJest(
    {
      ABQ_SOCKET: socketString,
    },
    async () => {
      const serverMessages = (await getMessages()).map(
        filterTestResultForSnapshot,
      );

      expect(serverMessages).toHaveLength(1);
      const results = serverMessages[0];
      expect(results).toHaveLength(9);

      const ids = new Set();
      for (const result of results) {
        ids.add(result.id);
      }

      expect(ids.size).toBe(9);

      expect(serverMessages).toMatchSnapshot();
    },
  );
});

ctest('ABQ mode runs focused tests', async () => {
  expect.assertions(4);

  const looped = pathForAbqTestFile('looped.test.js');
  const loopedRelative = path.relative(
    path.resolve(__dirname, '../abq'),
    looped,
  );

  const [socketString, getMessages] = await spawnServer([
    {
      focus: {
        test_ids: [
          `${loopedRelative}#2:1:0`,
          `${loopedRelative}#1:2:0`,
          `${loopedRelative}#0:0:0`,
        ],
      },
      id: looped,
      meta: {
        fileName: looped,
      },
      tags: [],
      type: 'test',
    },
  ]);

  await runAbqJest(
    {
      ABQ_SOCKET: socketString,
    },
    async () => {
      const serverMessages = (await getMessages()).map(
        filterTestResultForSnapshot,
      );

      expect(serverMessages).toHaveLength(1);
      const results = serverMessages[0];
      expect(results).toHaveLength(3);

      const ids = new Set();
      for (const result of results) {
        ids.add(result.id);
      }

      expect(ids.size).toBe(3);

      expect(serverMessages).toMatchSnapshot();
    },
  );
});
