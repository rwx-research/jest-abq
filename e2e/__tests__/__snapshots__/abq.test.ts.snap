// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ABQ mode handles ID generation for tests in loops 1`] = `
Array [
  Array [
    Object {
      "display_name": "looper i == i",
      "id": "__tests__/looped.test.js#0:0:0",
      "lineage": Array [
        "looper",
      ],
      "location": Object {
        "column": 7,
        "file": "<<REPLACED>>/e2e/abq/__tests__/looped.test.js",
        "line": 12,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
    Object {
      "display_name": "looper i == i",
      "id": "__tests__/looped.test.js#1:0:0",
      "lineage": Array [
        "looper",
      ],
      "location": Object {
        "column": 7,
        "file": "<<REPLACED>>/e2e/abq/__tests__/looped.test.js",
        "line": 12,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
    Object {
      "display_name": "looper i == i",
      "id": "__tests__/looped.test.js#2:0:0",
      "lineage": Array [
        "looper",
      ],
      "location": Object {
        "column": 7,
        "file": "<<REPLACED>>/e2e/abq/__tests__/looped.test.js",
        "line": 12,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
    Object {
      "display_name": "looper i == i",
      "id": "__tests__/looped.test.js#0:1:0",
      "lineage": Array [
        "looper",
      ],
      "location": Object {
        "column": 7,
        "file": "<<REPLACED>>/e2e/abq/__tests__/looped.test.js",
        "line": 12,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
    Object {
      "display_name": "looper i == i",
      "id": "__tests__/looped.test.js#1:1:0",
      "lineage": Array [
        "looper",
      ],
      "location": Object {
        "column": 7,
        "file": "<<REPLACED>>/e2e/abq/__tests__/looped.test.js",
        "line": 12,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
    Object {
      "display_name": "looper i == i",
      "id": "__tests__/looped.test.js#2:1:0",
      "lineage": Array [
        "looper",
      ],
      "location": Object {
        "column": 7,
        "file": "<<REPLACED>>/e2e/abq/__tests__/looped.test.js",
        "line": 12,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
    Object {
      "display_name": "looper i == i",
      "id": "__tests__/looped.test.js#0:2:0",
      "lineage": Array [
        "looper",
      ],
      "location": Object {
        "column": 7,
        "file": "<<REPLACED>>/e2e/abq/__tests__/looped.test.js",
        "line": 12,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
    Object {
      "display_name": "looper i == i",
      "id": "__tests__/looped.test.js#1:2:0",
      "lineage": Array [
        "looper",
      ],
      "location": Object {
        "column": 7,
        "file": "<<REPLACED>>/e2e/abq/__tests__/looped.test.js",
        "line": 12,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
    Object {
      "display_name": "looper i == i",
      "id": "__tests__/looped.test.js#2:2:0",
      "lineage": Array [
        "looper",
      ],
      "location": Object {
        "column": 7,
        "file": "<<REPLACED>>/e2e/abq/__tests__/looped.test.js",
        "line": 12,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
  ],
]
`;

exports[`ABQ mode handles errors in a test 1`] = `
Array [
  Array [
    Object {
      "display_name": "this test is erroring",
      "id": "__tests__/errors.test.js#0:0",
      "lineage": Array [],
      "location": Object {
        "column": 1,
        "file": "<<REPLACED>>/e2e/abq/__tests__/errors.test.js",
        "line": 9,
      },
      "meta": Object {},
      "output": "  ● this test is erroring

    can't catch me

       8 |
       9 | test('this test is erroring', () => {
    > 10 |   throw new Error("can't catch me");
         |         ^
      11 | });
      12 |

      at Object.<anonymous> (__tests__/errors.test.js:10:9)
",
      "runtime": 12,
      "status": Object {
        "backtrace": Array [
          "at Object.<anonymous> (<<REPLACED>>/e2e/abq/__tests__/errors.test.js:10:9)",
        ],
        "exception": "can't catch me",
        "type": "failure",
      },
    },
  ],
]
`;

exports[`ABQ mode handles errors outside of test execution 1`] = `
Array [
  Array [
    Object {
      "display_name": "<<REPLACED>>/e2e/abq/__tests__/error_setup.test.js",
      "id": "<<REPLACED>>/e2e/abq/__tests__/error_setup.test.js",
      "meta": Object {},
      "output": "  ●  can't catch me

       7 |  */
       8 |
    >  9 | throw new Error("can't catch me");
         |       ^
      10 |
      11 | // eslint-disable-next-line no-unreachable
      12 | test('true', () => {

      at Object.<anonymous> (__tests__/error_setup.test.js:9:7)
",
      "runtime": 12,
      "status": Object {
        "backtrace": Array [
          "Error: can't catch me",
          "    at Object.<anonymous> (<<REPLACED>>/e2e/abq/__tests__/error_setup.test.js:9:7)",
          "    at Runtime._execModule (<<REPLACED>>/packages/jest-runtime/build/index.js:1532:24)",
          "    at Runtime._loadModule (<<REPLACED>>/packages/jest-runtime/build/index.js:1115:12)",
          "    at Runtime.requireModule (<<REPLACED>>/packages/jest-runtime/build/index.js:975:12)",
          "<<TEST_RUNNER>>",
          "<<NODE INTERNAL>>",
          "    at runTestInternal (<<REPLACED>>/packages/jest-runner/build/runTest.js:287:16)",
          "    at runTest (<<REPLACED>>/packages/jest-runner/build/runTest.js:347:7)",
          "    at <<REPLACED>>/packages/jest-runner/build/index.js:215:11",
        ],
        "exception": "can't catch me",
        "type": "error",
      },
    },
  ],
]
`;

exports[`ABQ mode handles skipped tests 1`] = `
Array [
  Array [
    Object {
      "display_name": "one skipped test",
      "id": "__tests__/skip.test.js#0:0",
      "lineage": Array [],
      "location": Object {
        "column": 6,
        "file": "<<REPLACED>>/e2e/abq/__tests__/skip.test.js",
        "line": 9,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "pending",
      },
    },
    Object {
      "display_name": "other skipped test",
      "id": "__tests__/skip.test.js#1:0",
      "lineage": Array [],
      "location": Object {
        "column": 6,
        "file": "<<REPLACED>>/e2e/abq/__tests__/skip.test.js",
        "line": 11,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "pending",
      },
    },
  ],
]
`;

exports[`ABQ mode handles todo tests 1`] = `
Array [
  Array [
    Object {
      "display_name": "one todo test",
      "id": "__tests__/todo.test.js#0:0",
      "lineage": Array [],
      "location": Object {
        "column": 6,
        "file": "<<REPLACED>>/e2e/abq/__tests__/todo.test.js",
        "line": 9,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "todo",
      },
    },
    Object {
      "display_name": "other todo test",
      "id": "__tests__/todo.test.js#1:0",
      "lineage": Array [],
      "location": Object {
        "column": 6,
        "file": "<<REPLACED>>/e2e/abq/__tests__/todo.test.js",
        "line": 11,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "todo",
      },
    },
  ],
]
`;

exports[`ABQ mode runs focused tests 1`] = `
Array [
  Array [
    Object {
      "display_name": "looper i == i",
      "id": "__tests__/looped.test.js#0:0:0",
      "lineage": Array [
        "looper",
      ],
      "location": Object {
        "column": 7,
        "file": "<<REPLACED>>/e2e/abq/__tests__/looped.test.js",
        "line": 12,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
    Object {
      "display_name": "looper i == i",
      "id": "__tests__/looped.test.js#2:1:0",
      "lineage": Array [
        "looper",
      ],
      "location": Object {
        "column": 7,
        "file": "<<REPLACED>>/e2e/abq/__tests__/looped.test.js",
        "line": 12,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
    Object {
      "display_name": "looper i == i",
      "id": "__tests__/looped.test.js#1:2:0",
      "lineage": Array [
        "looper",
      ],
      "location": Object {
        "column": 7,
        "file": "<<REPLACED>>/e2e/abq/__tests__/looped.test.js",
        "line": 12,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
  ],
]
`;

exports[`ABQ_GENERATE_MANIFEST sends the manifest to the socket 1`] = `
Array [
  Object {
    "manifest": Object {
      "init_meta": Object {},
      "members": Array [
        Object {
          "id": "__tests__/equals.test.js",
          "meta": Object {
            "fileName": "__tests__/equals.test.js",
          },
          "tags": Array [],
          "type": "test",
        },
        Object {
          "id": "__tests__/error_setup.test.js",
          "meta": Object {
            "fileName": "__tests__/error_setup.test.js",
          },
          "tags": Array [],
          "type": "test",
        },
        Object {
          "id": "__tests__/errors.test.js",
          "meta": Object {
            "fileName": "__tests__/errors.test.js",
          },
          "tags": Array [],
          "type": "test",
        },
        Object {
          "id": "__tests__/failing.test.js",
          "meta": Object {
            "fileName": "__tests__/failing.test.js",
          },
          "tags": Array [],
          "type": "test",
        },
        Object {
          "id": "__tests__/looped.test.js",
          "meta": Object {
            "fileName": "__tests__/looped.test.js",
          },
          "tags": Array [],
          "type": "test",
        },
        Object {
          "id": "__tests__/skip.test.js",
          "meta": Object {
            "fileName": "__tests__/skip.test.js",
          },
          "tags": Array [],
          "type": "test",
        },
        Object {
          "id": "__tests__/sum.test.js",
          "meta": Object {
            "fileName": "__tests__/sum.test.js",
          },
          "tags": Array [],
          "type": "test",
        },
        Object {
          "id": "__tests__/todo.test.js",
          "meta": Object {
            "fileName": "__tests__/todo.test.js",
          },
          "tags": Array [],
          "type": "test",
        },
      ],
    },
  },
]
`;

exports[`ABQ_SOCKET runs Jest in ABQ mode 1`] = `
Array [
  Array [
    Object {
      "display_name": "this test is failing",
      "id": "__tests__/failing.test.js#0:0",
      "lineage": Array [],
      "location": Object {
        "column": 1,
        "file": "<<REPLACED>>/e2e/abq/__tests__/failing.test.js",
        "line": 9,
      },
      "meta": Object {},
      "output": "  ● this test is failing

    expect(received).toBe(expected) // Object.is equality

    Expected: false
    Received: true

       8 |
       9 | test('this test is failing', () => {
    > 10 |   expect(true).toBe(false);
         |                ^
      11 | });
      12 |
      13 | test('this test is also failing', () => {

      at Object.toBe (__tests__/failing.test.js:10:16)
",
      "runtime": 12,
      "status": Object {
        "backtrace": Array [
          "at Object.toBe (<<REPLACED>>/e2e/abq/__tests__/failing.test.js:10:16)",
        ],
        "exception": "expect(received).toBe(expected) // Object.is equality

Expected: false
Received: true",
        "type": "failure",
      },
    },
    Object {
      "display_name": "this test is also failing",
      "id": "__tests__/failing.test.js#1:0",
      "lineage": Array [],
      "location": Object {
        "column": 1,
        "file": "<<REPLACED>>/e2e/abq/__tests__/failing.test.js",
        "line": 13,
      },
      "meta": Object {},
      "output": "  ● this test is also failing

    expect(received).toBe(expected) // Object.is equality

    Expected: false
    Received: true

      12 |
      13 | test('this test is also failing', () => {
    > 14 |   expect(true).toBe(false);
         |                ^
      15 | });
      16 |

      at Object.toBe (__tests__/failing.test.js:14:16)
",
      "runtime": 12,
      "status": Object {
        "backtrace": Array [
          "at Object.toBe (<<REPLACED>>/e2e/abq/__tests__/failing.test.js:14:16)",
        ],
        "exception": "expect(received).toBe(expected) // Object.is equality

Expected: false
Received: true",
        "type": "failure",
      },
    },
  ],
  Array [
    Object {
      "display_name": "the + operator 1 + 2 equals 3",
      "id": "__tests__/sum.test.js#0:0:0",
      "lineage": Array [
        "the + operator",
      ],
      "location": Object {
        "column": 3,
        "file": "<<REPLACED>>/e2e/abq/__tests__/sum.test.js",
        "line": 10,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
    Object {
      "display_name": "the + operator 2 + 3 equals 5",
      "id": "__tests__/sum.test.js#1:0:0",
      "lineage": Array [
        "the + operator",
      ],
      "location": Object {
        "column": 3,
        "file": "<<REPLACED>>/e2e/abq/__tests__/sum.test.js",
        "line": 14,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
    Object {
      "display_name": "the + operator with three operands 2 + 3 + 4 equals 9",
      "id": "__tests__/sum.test.js#0:2:0:0",
      "lineage": Array [
        "the + operator",
        "with three operands",
      ],
      "location": Object {
        "column": 5,
        "file": "<<REPLACED>>/e2e/abq/__tests__/sum.test.js",
        "line": 19,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
  ],
]
`;

exports[`Reports all tests in a file 1`] = `
Array [
  Array [
    Object {
      "display_name": "the + operator 1 + 2 equals 3",
      "id": "__tests__/sum.test.js#0:0:0",
      "lineage": Array [
        "the + operator",
      ],
      "location": Object {
        "column": 3,
        "file": "<<REPLACED>>/e2e/abq/__tests__/sum.test.js",
        "line": 10,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
    Object {
      "display_name": "the + operator 2 + 3 equals 5",
      "id": "__tests__/sum.test.js#1:0:0",
      "lineage": Array [
        "the + operator",
      ],
      "location": Object {
        "column": 3,
        "file": "<<REPLACED>>/e2e/abq/__tests__/sum.test.js",
        "line": 14,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
    Object {
      "display_name": "the + operator with three operands 2 + 3 + 4 equals 9",
      "id": "__tests__/sum.test.js#0:2:0:0",
      "lineage": Array [
        "the + operator",
        "with three operands",
      ],
      "location": Object {
        "column": 5,
        "file": "<<REPLACED>>/e2e/abq/__tests__/sum.test.js",
        "line": 19,
      },
      "meta": Object {},
      "output": null,
      "runtime": 12,
      "status": Object {
        "type": "success",
      },
    },
  ],
]
`;
